File: gitdot-auth/src/handler/device/authorize_device.rs
use axum::{Json, extract::State, http::StatusCode};

use gitdot_api::endpoint::auth::device::authorize_device as api;
use gitdot_axum::Principal;
use gitdot_core::dto::AuthorizeDeviceRequest;

use crate::app::{AppError, AppResponse, AppState};

pub async fn authorize_device(
    principal: Principal,
    State(state): State<AppState>,
    Json(body): Json<api::AuthorizeDeviceRequest>,
) -> Result<AppResponse<()>, AppError> {
    let request = AuthorizeDeviceRequest::new(&body.user_code, principal.id)?;
    state
        .device_service
        .authorize_device(request)
        .await
        .map_err(AppError::from)
        .map(|_| AppResponse::new(StatusCode::OK, ()))
}

File: gitdot-web/packages/gitdot-api/src/endpoint/auth/device/authorize-device.ts
import { z } from "zod";
import type { Endpoint } from "../../endpoint";

export const AuthorizeDeviceRequest = z.object({
  user_code: z.string(),
});
export type AuthorizeDeviceRequest = z.infer<typeof AuthorizeDeviceRequest>;

export const AuthorizeDeviceResponse = z.void();
export type AuthorizeDeviceResponse = z.infer<typeof AuthorizeDeviceResponse>;

export const AuthorizeDevice = {
  path: "/auth/device/authorize",
  method: "POST",
  request: AuthorizeDeviceRequest,
  response: AuthorizeDeviceResponse,
} as const satisfies Endpoint;

export type AuthorizeDevice = typeof AuthorizeDevice;
